### ISSUE TITLE
COBOL - Folding problem on an "if" / "end-if" block when there is a comment line inside the block 

### ISSUE THREAD

**Issue by FALLAI-Denis** [IBM/zopeneditor-about#101]
## Development environment used

- Z Open Editor version:
- Editor Platform
  - [X] Visual Studio Code
  - [ ] CodeReady Workspaces
  - [ ] Eclipse Che
  - [ ] Standalone Theia
- Editor Platform Version: 1.51.1
- Operating System: Windows 10 1607 :-(
- Java Version: java version "1.8.0_102" / Java(TM) SE Runtime Environment (build 1.8.0_102-b14) / Java HotSpot(TM) 64-Bit Server VM (build 25.102-b14, mixed mode)
- [ ] Related to RSE API?
  - RSE API Plugin version:
  - Zowe CLI version: 6.24.5
  - Node.js version: v12.16.2
- Logs attached: no

## Problem Description

Detailed steps for reproducing the problem:

1. Create an "if" / "end-if" block
2. Insert a comment line ("*" in column 7) inside the block
3. Test the folding on the "if" instruction

### Observed behavior

The folding stops at the comment line instead of extending to the "end-if"

![Folding with comments](https://user-images.githubusercontent.com/63861965/100267621-03a78280-2f54-11eb-8314-fd7e0895e2bb.gif)

### Expected behavior

The folding must extend to the "end-if".

**Comment by FALLAI-Denis** [comment]
@phaumer 
Not a defect ?

**Comment by phaumer** [comment]
@FALLAI-Denis no, as we have not provided any folding features at all so far. The folding you see is VS Code itself doing some guessing based on indentation and so on. There are VS Code APIs to control folding, but we have not looked into them, yet. However, I personally agree that this would be a great enhancement to get the behavior right and fully support at some point. So all the input you are providing around this with your requirements are really valuable to us.

**Comment by FALLAI-Denis** [comment]
A possible solution is described here #141

**Comment by FALLAI-Denis** [comment]
Hi,

Could this feature be supported by the folding solution integrated in ZOE 4.4.0?

By also adding the management of the different COBOL instructions supporting an end-of-scope instruction, mainly:
* if / end-if
* evaluate / end-evaluate
* perform / end-perform (if it is possible to distinguish perform inline from perform paragraph / section)

A priori this functionality can only be handled by an analysis of the abstract syntax tree and not by a simple analysis of the source code lines as is done in the solution described in discussion #111.

Thanks.

**Comment by phaumer** [comment]
@FALLAI-Denis perhaps I misunderstand you request, but can you show an example where this does not work in 4.4.0? I did this:

![2024-08-13_08-52-06 (1)](https://github.com/user-attachments/assets/74ac063c-b6c6-47e6-97ff-b0cc2b7f4975)

**Comment by FALLAI-Denis** [comment]
Hi @phaumer,

Actually, as this feature was not explicitly announced in the changelog, I did not test it 😳... sorry.
I see that the online documentation details the [COBOL folding functions](https://ibm.github.io/zopeneditor-about/Docs/editor_codechanges.html#cobol-code-folding).

I confirm that folding works on `IF` and `WHEN`, but also, ie, on `AT END` (`READ` statement).
On the other hand, there is no folding on `ELSE`, (despite the indications in the online documentation), it is in fact a folding on the _block_ of statements after the `ELSE` which is proposed.
It has a fold proposed on the `EVALUATE` _block_, but not on the `PERFORM` inline _block_.

The folding on the `IF` is ambiguous: is this a fold on the `IF` _block_ or a fold on the `IF` _condition_?
* if it is a folding on the _condition_, then the folding should stop before the `ELSE` of the same level, which is not the case,
* if the folding is on the `IF` _block_, then this explains that it encompasses the `ELSE` of the same level, but then it is difficult to position oneself on the `ELSE` by a folding at the level of the `IF`.

Personally I would have preferred a _block_ approach, (Which is the case for `EVALUATE`, `EXEC / END-EXEC` blocks, Sections and Paragraphs: they are blocks).

Maybe a combination of both is needed:
* folding on `IF` (until `ELSE` if present), `ELSE` (currently missing), `WHEN` _conditions_, in fact all conditional parts of a statement,
* folding on `EVALUATE`, `PERFORM` inline _blocks_, in fact all statements which can include one or more _block_ of dependant statements , (and which generally have an end-of-scope statement `END-xxx`).

But there will be a problem when the same line meets multiple folding criteria, (ie `EVALUATE` and `WHEN` on the same source line).
Who should be given priority? I think in this case that we should prioritize folding over the _condition_; (which therefore cancels folding on the _block_).
This is the case for the `IF` statement which is both a _block_ and a _condition_, in this case, favor/prioritize folding on the _condition_ and abandon folding on the _block_.

I also noted folding proposals in the following cases:
* statement written on several source lines
> * including `DATA-ITEM` declarations, (in the introduction to the online documentation I think we are referring more to folding on the `GROUP DATA-ITEMS`?)
> * including `PERFORM` when written on several lines in case of _condition_ or `THRU`, (At first I thought the folding was due to the presence of a _condition_ in the case of an `PERFORM` inline, which is misleading)
* block of conditionned lines, (maybe what is called `DO` in the online documentation?) 

--- 

I had random behavior on some constructions, in the same program, on the same statements:
* sometimes the `EVALUATE` _block_ is collapsible, sometimes it is not...
* sometimes the `WHEN OTHER` _block_ is collapsible, sometimes it is not...

I am not able to define the conditions of the problem nor reproduce it...
This may have distorted my interpretations.

Folding a statements _block_ (`DO`) that comes right after folding a _condition_ seems a bit redundant to me... 
There is a folding conflict between folding on a _block_ of statements and folding on an statement written on several lines, if the first statement of the _block_ is written on several lines, (see the case on the program SAM1.cbl, lines 361 to 367, or lines 368 to 374).
This would argue against proposing folding on a _block_ already dependent on a _condition_.
--> Do you want me to open a feature request issue for this?

The benefit of folding on all instructions written on multiple lines seems limited to me, and it would be preferable to introduce folding on all instructions that can handle dependent instruction blocks.
--> Do you want me to open a feature request issue for this?

The folding feature apparently does not work on copybooks, is there a reason for this limitation? (the fact that the source code of a copybook is not necessarily a complete code?)
--> Do you want me to open a feature request issue for this?

There is an error in the management of `WHEN` not followed by an imperative statement, (or'ed `WHEN`): the folding on the first `WHEN` goes too far, (see use case on SAM1.cbl program, lines 428: 2 or'ed `WHEN`).
--> Do you want me to open a bug issue for this?

Moreover, during my tests, the folding feature stopped working, after switching to copybook source and returning to the original COBOL source.
--> Do you want me to open a bug issue for this?

Thanks.

**Comment by phaumer** [comment]
@FALLAI-Denis thanks so much for the very detailed review on this one. No new Github issues are required as we track this in-house in one user story and I collected all your previous issues and these comments now. We will go through them all and discuss with other stakeholders for the next round of refinements.
